### PR TITLE
devpkg: better flake references and installable parsing

### PR DESCRIPTION
**Pull request by gcurtis**
In anticipation of adding support for non-default package outputs, I wanted to clean up some of the flakeref parsing in `devpkg`. This will make it easier to handle Nix's `flakeref#attrpath^outputs` installable syntax. This PR is broken up into two commits: one that handles flakeref parsing and one that handles installable parsing.

The `ParseFlakeRef` function is able to parse the more common flakerefs that devbox currently supports ("path", "flake", "indirect", etc.). It doesn't support mercurial, gitlab, and sourcehuthut for now. The `ParseFlakeInstallable` function parses Nix flake installables, which are a superset of the flake reference syntax. It allows an attribute path and/or outputs to be specified using the `flake#attrpath^outputs` syntax. The tests were validated using the nix expression `builtins.parseFlakeRef "ref"` and the `nix flake metadata <ref>` CLI.

The goal isn't to perfectly replicate Nix's parsing of flake references (which itself is sometimes inconsistent), but to make it good enough to handle flakerefs in devbox.json.

### CODE REVIEW

File: internal/devpkg/flakeref.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
+package devpkg
+
+import (
+	"net/url"
+	"strings"
+
+	"go.jetpack.io/devbox/internal/redact"
+)
+
+// FlakeRef is a parsed Nix flake reference. A flake reference is as subset of
```
Comment (savil) [inline internal/devpkg/flakeref.go:10]: `...is a subset of...`

File: internal/devpkg/flakeref.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+	// "github".
+	Owner string `json:"owner,omitempty"`
+	Repo  string `json:"repo,omitempty"`
+
+	// Rev and ref are the git revision (commit hash) and ref
+	// (branch or tag) when Type is "github" or "git".
+	Rev string `json:"rev,omitempty"`
+	Ref string `json:"ref,omitempty"`
+
+	// Dir is non-empty when the directory containinig the flake.nix file is
```
Comment (savil) [inline internal/devpkg/flakeref.go:41]: typo: containing

File: internal/devpkg/flakeref.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+		return parsed, nil
+	}
+	parsed, _, err := parseFlakeURLRef(ref)
+	return parsed, err
+}
+
+func parseFlakeURLRef(ref string) (parsed FlakeRef, fragment string, err error) {
+	// A good way to test how Nix parses a flake reference is to run:
+	//
+	// 	nix eval --json --expr 'builtins.parseFlakeRef "ref"' | jq
```
Comment (savil) [inline internal/devpkg/flakeref.go:101]: good pro-tip 👍🏾

**Review by savil — APPROVED** [review]
This is great! What is the plan to integrate it with the rest of the code?

> I wanted to clean up some of the flakeref parsing in devpkg. This will make it easier to handle Nix's flakeref#attrpath^outputs installable syntax. This PR is broken up into two commits: one that handles flakeref parsing and one that handles installable parsing.

Is the expectation that users would directly run `devbox add flakeref#attrpath^outputs`? If so, I worry that is a complicated string for our users. We can find a simpler way.

On the other hand, if we're saying the user's commands will get translated by Devbox CLI code into being of the form `flakeref#attrpath^outputs` and then this makes sense to me.

This product discussion is probably better in a different forum than this PR. Happy to chat elsewhere more :)

**Comment by gcurtis** [comment]
It's the second one. For handling the flake-like syntax supported in devbox.json (which users have to add manually).

**Pull request merged**